apps/emqx_dashboard/src/emqx_dashboard_admin.erl

```diff
@@ -115,18 +114,31 @@ add_default_user() ->
 %% API
 %%--------------------------------------------------------------------
 
--spec add_user(dashboard_username(), binary(), dashboard_user_role(), binary()) ->
+-spec add_user(
+    dashboard_username(), _Password :: binary(), dashboard_user_role(), _Desc :: binary()
+) ->
     {ok, map()} | {error, any()}.
-add_user(Username, Password, Role, Desc) when is_binary(Username), is_binary(Password) ->
-    case {legal_username(Username), legal_password(Password), legal_role(Role)} of
-        {ok, ok, ok} -> do_add_user(Username, Password, Role, Desc);
-        {{error, Reason}, _, _} -> {error, Reason};
-        {_, {error, Reason}, _} -> {error, Reason};
-        {_, _, {error, Reason}} -> {error, Reason}
+add_user(Username, Password, Role0, Desc) when is_binary(Username), is_binary(Password) ->
+    maybe
+        ok ?= legal_username(Username),
+        ok ?= legal_password(Password),
+        {ok, ParsedRole} ?= parse_role(Role0),
+        #{?role := Role} = ParsedRole,
+        Extra = parsed_role_to_extra(ParsedRole),
+        do_add_user(Username, Password, Role, Desc, Extra)
     end.
 
-do_add_user(Username, Password, Role, Desc) ->
-    Res = mria:sync_transaction(?DASHBOARD_SHARD, fun add_user_/4, [Username, Password, Role, Desc]),
+parsed_role_to_extra(#{?namespace := Ns}) when is_binary(Ns) ->
+    #{?namespace => Ns};
```

Comment:
Q: we need `password_ts` or not?

Reply:
`add_user_/5` in this module takes care of injecting that attribute.